[FEATURE#35371] Provider de traduction Abort Api

rinaz-a opened the pull request:
On crée ce petit provider afin de traduire les messages abort de l'api dans le language de l'utilisateur, le reste est dans le readMe :)

rinaz-a committed:
[FEATURE#35371] Provider de traduction Abort Api

On crée ce petit provider afin de traduire les messages abort de l'api dans le language de l'utilisateur, le reste est dans le readMe :)

TinyTinyKame commented on README.md:
```diff
@@ -0,0 +1,40 @@
+# composer-translation-service-provider
+Translation service provider
+
+Ce provider nous permettre de revoyer les messages `abort` dans la langue de l'utilisateur à partir de nos API vers nos clients.
```
`+Ce provider va nous permettre de renvoyer les messages`abort`dans la langue de l'utilisateur à partir de nos API vers nos clients.`

TinyTinyKame commented on README.md:
```diff
@@ -0,0 +1,40 @@
+
+Ce provider nous permettre de revoyer les messages `abort` dans la langue de l'utilisateur à partir de nos API vers nos clients.
+
+Il faut ajouter un (ou plusieurs) fichier de traduction et les ajouter au EtnaConfig.php à la façon suivante :
```
`de la façon`

TinyTinyKame commented on README.md:
````diff
@@ -0,0 +1,40 @@
+
+```
+    return $app->abort(400, "Bad request : operation forbidden on a deleted %entity%",[
+        'Super-Translation' => json_encode(['%entity%' => $entity_name])
````
`Super-Translation`? genre, t'as pas trouvé un autre nom de header :p

TinyTinyKame commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,39 @@
+        $app->after(
+            function (Request $request, Response $response) use ($app) {
+                //on récupère les tokens que l'on a surchargé dans la réponse
+                $tokens = json_decode($response->headers->get('Super-Translation', "[]"), true);
```
un autre nom que `Super-Translation` quand même

TinyTinyKame reviewed (changes requested):
J'ai une autre question du coup, est-ce que les headers que tu rajoutes disparaissent lorsque le réponse est envoyée?

Leosten commented on README.md:
````diff
@@ -0,0 +1,40 @@
+</xliff>
+```
+
+`%name%` étant ici un token (optionel) que l'on peut ajouter à la traduction, il faudra le placer dans le header du `abort` comme ceci :
````
`(optionnel)`

gdubost commented on composer.json:
```diff
@@ -0,0 +1,34 @@
+    "require": {
+        "php": ">=5.5",
+        "silex/silex": "~2.0@dev",
+        "dflydev/doctrine-orm-service-provider": "2.x"
```
pourquoi ?

rinaz-a replied:
whaaat je l'ai enlevé chez moi, chelou

gdubost commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,39 @@
+                $tokens = json_decode($response->headers->get('Super-Translation', "[]"), true);
+
+                // on récupère le contenu de la réponse càd le message d'erreur
+                $content = json_decode($response->getContent(), TRUE);
```
les minuscules c'est bien aussi

gdubost commented:
Mais du coup il manque deux trois trucs, faudrait gérer la langue principale du navigateur et rajouter des tests ^^

lapwil commented on README.md:
````diff
@@ -0,0 +1,40 @@
+```
+
+`%name%` étant ici un token (optionel) que l'on peut ajouter à la traduction, il faudra le placer dans le header du `abort` comme ceci :
+/!\ Le nom du token ne doit pas contenir de caractères spéciaux, on préfèrera ici le camelCase.
````
dommage que ton exemple ne reprennes pas cette idée de camelCase 🐫  tu pourrais choisir `%entityName%` plutôt

lapwil commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,39 @@
+
+namespace ETNA\Silex\Provider\Translation;
+
+use Silex\Provider\DoctrineServiceProvider;
```
no need for this

rinaz-a commented:
les tests sont sur l'API @gdubost

rinaz-a committed:
[FEATURE#35371] On ajoute le check de la locale de l'utilisateur et on supprime la surcharge des headers

lapwil commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,42 @@
+            function (Request $request, Response $response) use ($app) {
+                if ('fr' === $request->getLocale()) {
+                    //on récupère les tokens que l'on a surchargé dans la réponse
+                    $tokens = json_decode($response->headers->get('Super-Translation', "[]"), true);
```
comme disait @TinyTinyKame tu veux pas faire quelque chose d'autre que `Super-Translation` genre `Api-Translation` ou un truc plus explicite 😛

rinaz-a replied:
étrangement je pensais l'avoir edit mais j'ai dut edit celui que j'ai pour mes tests dans contract-api, my bad

lapwil reviewed (approved):
OK quand yaura les tests

rinaz-a committed:
[FEATURE#35371] Tests Added + totalement fonctionnels coverage compris

- boostrap correct de behat, composer, drone
- ajout du readMe
- boostrap d'une app pour les tests
- on test si l'utilisateur est français ou non
- test et gestion des tokens
- ExceptionListener pour gerer l'app comme on le fait sur les divers api

Leosten commented on app/Utils/Silex/Translator/abort.fr.xlf:
```diff
@@ -0,0 +1,15 @@
+            </trans-unit>
+            <trans-unit id="2">
+                <source>Manager not found : %id%</source>
+                <target>Résponsable non trouvé : %id%</target>
```
`Responsable`

staciekith commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,59 @@
+                }
+
+                // on clean le header que l'on a ajouté
+                $response->headers->remove('Translation');
```
Ca bug pas si on n'a pas le header Translation ?

rinaz-a replied:
Non ça ne bug pas mais je vais le mettre dans le isset cat ça ne sert à rien de faire appel à cette fonction quand on en a pas besoin ;)

staciekith commented on Tests/Bootstrap/FeatureContext.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * @Then /^je devrais avoir un résultat d\'API en JSON$/
+     */
+    public function jeDevraisAvoirUnResultatDApiEnJSON()
```
Ces fonctions sont déjà dans le composer-behat-utils non ?

rinaz-a replied:
Oui mais elles nécéssitent une base de test dont je n'ai pas besoin si on appelle ApiContext && DoctrineContext du coup je les ai un peu mod pour moins ajouter de trucs useless

gdubost replied:
On a fait un composer pour éviter ce genre de choses, surtout que le doctrine context a priori t'es pas sense l'utiliser ^^

Hazainn replied:
@rinaz-a: du coup tu as fais quelque chose pour ce commentaire ?

rinaz-a replied:
Merci pour l'iddentification, j'étais pas remonté, non je touche pas car dans le `ApiContext` encore une fois on appelle `DoctrineContext` et que je n'en veux pas : 
```
    /**
     * @When /^je fais un (GET|POST|PUT|DELETE|OPTIONS) sur ((?:[a-zA-Z0-9,:!\/\.\?\&\=\+_%-]*)|"(?:[^"]+)") avec le JSON suivant :$/
     */
    public function jeFaisUneRequetteHTTPAvecDuJSON($method, $url, $body)
    {
        if (preg_match('/^".*"$/', $url)) {
            $url = substr($url, 1, -1);
        }

        if ($body !== null) {
            if (is_object($body)) {
                $body = $body->getRaw();
            }
            $this->request["headers"]["Content-Type"] = 'application/json';
            // add content-length ...
        }

        if (null === $body && isset($this->request["headers"]["Content-Type"])) {
            unset($this->request["headers"]["Content-Type"]);
        }

        $request = Request::create($this->base_url . $url, $method, [], [], [], [], $body);
        $request->headers->add($this->request["headers"]);
        $request->cookies->add($this->request["cookies"]);
        $request->files->add($this->request["files"]);

        $response = self::$silex_app->handle($request, HttpKernelInterface::MASTER_REQUEST, true);

        $result = [
            "http_code"    => $response->getStatusCode(),
            "http_message" => Response::$statusTexts[$response->getStatusCode()],
            "body"         => $response->getContent(),
            "headers"      => array_map(
                function ($item) {
                    return $item[0];
                },
                $response->headers->all()
            ),
        ];

        $this->getContext('ETNA\FeatureContext\DoctrineContext')->checkMaxQueries($method, $result);

        $this->response = $result;
    }
```

staciekith commented on composer.json:
```diff
@@ -0,0 +1,57 @@
+    },
+    "require-dev": {
+        "satooshi/php-coveralls": "dev-master",
+        "etna/composer-behat-utils": "2.x-dev",
```
C'est required mais pas utilisé :p idem pour doctrine juste en dessous

rinaz-a replied:
Si c'est utilisé va voir dans le behat.yml j'utilise MainContext

rinaz-a replied:
Mais vrai pour doctrine je croyais l'avoir retiré mais y'en avait 3...

lapwil commented on Tests/Bootstrap/FeatureContext.php:
```diff
@@ -0,0 +1,154 @@
+     */
+    public function jeFaisUneRequetteHTTP($method, $url, $body = null)
+    {
+        if ($body !== null) {
```
ta condition est dans le mauvais sens!

exemple: `$variable === 'toto'` => `'toto' === $variable`

lapwil commented on Tests/Bootstrap/FeatureContext.php:
```diff
@@ -0,0 +1,154 @@
+            $url = substr($url, 1, -1);
+        }
+
+        if ($body !== null) {
```
ta condition est dans le mauvais sens!

exemple: `$variable === 'toto'` => `'toto' === $variable`

lapwil commented on Tests/Features/translation.feature:
```diff
@@ -0,0 +1,45 @@
+        "Bad request : operation forbidden on a deleted contract"
+    """
+
+@GET @wip
```
un wip sauvage

lapwil commented on app/Utils/Silex/EventSubscriber/ExceptionListener.php:
```diff
@@ -0,0 +1,61 @@
+                break;
+
+            default:
+                $code = 500;
```
ce n'est pas aligné ici

rinaz-a replied:
oui je n'avais pas commit le déplacement de la variable header #boulay mais c'est solved :)

lapwil reviewed (approved):
OK

TinyTinyKame commented on Tests/Features/translation.feature:
```diff
@@ -0,0 +1,45 @@
+    """
+
+@GET
+Scénario: On cherche à traduire un message après un app->abort
```
tous tes noms de scénarios sont les mêmes

gdubost commented on .sync.yml:
```diff
@@ -0,0 +1,7 @@
+    name: "translation-service-provider"
+    minor_version: "0.1"
+    badges:
+        - "[![Build Status](http://drone.etna-alternance.net/github.com/etna-alternance/translation-service-provider)]"
```
C'est quoi ce fichier ??

gdubost commented on behat.yml:
```diff
@@ -0,0 +1,43 @@
+                - ETNA\FeatureContext\MainContext
+                - ETNA\FeatureContext\FixedDateContext:
+                    date: "2016-04-12 14:42:42"
+                - ETNA\FeatureContext\AuthContext
```
Ah bon ?

rinaz-a replied:
bah c'est partout j'ai supposé que fallait le mettre pour des éventuels futurs tests mais c'est con de le load pour rien je vais l'enlever

gdubost commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,59 @@
+                //on récupère les tokens que l'on a surchargé dans la réponse
+                $tokens = json_decode($response->headers->get('Translation', "[]"), true);
+                if (isset($tokens["Translation"])) {
+                    $tokens = json_decode($tokens["Translation"], true);
```
Wat ?

Tes headers font donc `$headers['Translation']['Translation']` ??

rinaz-a replied:
non c'est bien $headers['Translation'] mais quand tu get ça te garde la clée

gdubost commented:
Ils sont passes ou les badges du readme ?

rinaz-a commented:
Je m'occupe de remonter les badges et modifier les noms de scénario ;)

misseur reviewed (approved):
A part Schrutinizer qui meurt à petit feu (3222 minutes qu'il mouline dans le vent 😄 ), ça me semble ok :)

gdubost commented on README.md:
```diff
@@ -0,0 +1,43 @@
+
+[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/etna-alternance/composer-translation-provider/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/etna-alternance/composer-translation-provider/?branch=master)
+[![Coverage Status](https://coveralls.io/repos/github/etna-alternance/composer-translation-provider/badge.svg?branch=master)](https://coveralls.io/github/etna-alternance/composer-translation-provider?branch=master)
+
```
Rajoute celui la :
```
[![Build Status](http://drone.etna-alternance.net/api/badge/github.com/etna-alternance/composer-translation-provider/status.svg?branch=master)](http://drone.etna-alternance.net/github.com/etna-alternance/composer-translation-provider)
```

gdubost commented on app/Utils/Silex/EventSubscriber/ExceptionListener.php:
```diff
@@ -0,0 +1,61 @@
+        if (is_a($exception, "InvalidArgumentException")) {
+            $code = 400;
+        }
+        if (is_a($exception, "Symfony\Component\HttpKernel\Exception\HttpException")) {
```
Si c'est nécessaire au bon fonctionnement de ton provider, ca doit en faire parti et être register par ton provider et pas seulement faire parti de l'app de test ^^

gdubost commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,59 @@
+                        }
+                    } elseif (isset($tokens)) {
+                        foreach ($tokens as $variable => $value) {
+                            $content = str_replace($variable, $value, $content);
```
Tu peux le faire en un seul appel a str_replace :)

Je t'invite a lire le man :D

TinyTinyKame commented on phpmd.xml:
```diff
@@ -0,0 +1,52 @@
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
+         xsi:noNamespaceSchemaLocation=" http://pmd.sf.net/ruleset_xml_schema.xsd">
+    <description> phpmd ruleset for studentmanager-api</description>
```
studentmanager-api

lapwil reviewed (approved):
Ca me semble correc'

TinyTinyKame commented on src/TranslationProvider.php:
```diff
@@ -0,0 +1,59 @@
+            $app->register(
+                new TranslationServiceProvider(),
+                [
+                    'locale_fallbacks' => ['fr']
```
hmmmm, je suis pas trop fan de ça en vrai, ça devrait fallback sur du `en` quand il n'y a pas la traduction `fr`

rinaz-a replied:
en fait c'est juste pour setter une locale si la requête n'en contient pas, pour moi comme on est en france ça devrait être fr, dans tous les cas si il n'y a pas de traduction fr le message est renvoyé en anglais, je veux bien changer mais de toutes façons c'est quasi impossible que la requête ne contient pas de locale, je l'ai mis plus par principe ce check

Hazainn commented on app/Utils/Silex/EventSubscriber/ExceptionListener.php:
```diff
@@ -0,0 +1,55 @@
+
+            default:
+                $code = 500;
+                break;
```
c'est pas forcément très utile les retours à la ligne après les break.

Hazainn commented on src/ExceptionListenerWithHeaders.php:
```diff
@@ -0,0 +1,61 @@
+            default:
+                $code = 500;
+                break;
+        }
```
ici aussi pour les espaces après les breaks.

Hazainn commented on src/ExceptionListenerWithHeaders.php:
```diff
@@ -0,0 +1,61 @@
+            KernelEvents::EXCEPTION => ['onKernelException', 32],
+        ];
+    }
+}
```
ce code ressemble énormément à celui-la : 
- `Utils/Silex/EventSubscriber/ExceptionListener.php`

il y a une raison ?

rinaz-a replied:
yep, pour que quoi qu'il arrive on greffe les headers à la response, qu'il y ai ou non un exceptionListener dans l'app de base ;)

rinaz-a committed:
[FEATURE#35371] Merci les maj, désormais la Translation de symfony remplace les tokens plus besoin de str_replace

TinyTinyKame reviewed (approved):
Okay

staciekith reviewed (approved):
okay % drone

gdubost commented on src/ExceptionListenerWithHeaders.php:
```diff
@@ -0,0 +1,59 @@
+                $code,
+                $headers
+            )
+        );
```
Ce listener fait trop de choses, je l'imaginais uniquement ajouter les headers à la réponse

rinaz-a replied:
j'ai limité au max là @gdubost

gdubost replied:
Même avec :
```
$response = $event->getResponse();
$response->setHeaders(...);
$event->setResponse($response);
```

lapwil commented on Tests/Bootstrap/FeatureContext.php:
```diff
@@ -0,0 +1,154 @@
+        if ("application/json" !== $this->response["headers"]["content-type"]) {
+            throw new \Exception("Invalid response type");
+        }
+        if ("" == $this->response['body']) {
```
tu pourrais la `trim()` histoire d'être sûr que c'est une chaîne "vide" et tu pourrais `===`

rinaz-a committed:
[FEATURE#35371] On n'ajoute plus d'ExceptionListener spécifique, on garde celui de l'app

lapwil reviewed (approved):
Ca me semble bien maintenant 😄

TinyTinyKame reviewed (approved):
Ça m'a l'air bon
